### PR TITLE
Upstream Authority integration

### PR DESCRIPTION
**Pull request by MarcosDY**
**Description of change**
Add nested-rotation integration test

**Which issue this PR fixes**
Fix: #1380

### CODE REVIEW

File: test/integration/suites/nested-rotation/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+           leafA-server            leafA-server  
+                |                       |
+           leafA-agent             leafA-agent             
+
+Test steps:
+
+- fetch x509-SVID from `intermediateA-agent` and validate them on `intermediateB-agent` using `spire-agent api validate x509` 
+- fetch x509-SVID from `leafA-agent` and validate them on `leafB-agent` using `spire-agent api validate x509` 
+- fetch jwt-SVID from `intermediateA-agent` and validate them on `intermediateB-agent` using `spire-agent api validate jwt` 
+- fetch jwt-SVID from `leafA-agent` and validate them on `leafB-agent` using `spire-agent api validate jwt`
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/README.md:26]:
```suggestion
- Fetch an X509-SVID from `intermediateA-agent` and it them on `intermediateB-agent`
- Fetch an X509-SVID from `leafA-agent` and validate it on `leafB-agent`
- Fetch a JWT-SVID from `intermediateA-agent` and validate it on `intermediateB-agent`
- Fetch a JWT-SVID from `leafA-agent` and validate it on `leafB-agent` 
```

File: test/integration/suites/nested-rotation/intermediateA/server/server.conf [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        }
+	}
+	KeyManager "memory" {
+		plugin_data = {}
+	}
+ 	UpstreamAuthority "spire" {
+ 	    plugin_data = {
+ 	        server_address      = "root-server"
+ 	        server_port         = 8081
+            workload_api_socket = "/opt/spire/sockets/workload_api.sock"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/intermediateA/server/server.conf:31]: weird indentation

File: test/integration/suites/nested-rotation/intermediateA/agent/agent.conf [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            directory = "/opt/spire/data/agent"
+        }
+    }
+    WorkloadAttestor "unix" {
+        plugin_data {
+        }
+    }
+    WorkloadAttestor "docker" {
+        plugin_data {
+            # Error when using default (1.41)
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/intermediateA/agent/agent.conf:29]: What was the error here? Is this a problem we need to address in SPIRE?

File: test/integration/suites/nested-rotation/intermediateB/server/server.conf [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        }
+	}
+	KeyManager "memory" {
+		plugin_data = {}
+	}
+ 	UpstreamAuthority "spire" {
+ 	    plugin_data = {
+ 	        server_address      = "root-server"
+ 	        server_port         = 8081
+            workload_api_socket = "/opt/spire/sockets/workload_api.sock"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/intermediateB/server/server.conf:31]: weird indentation

File: test/integration/suites/nested-rotation/leafA/server/server.conf [hunk 5]
```diff
@@ -0,0 +1,34 @@
+server {
+	bind_address = "0.0.0.0"
+	bind_port = "8081"
+	registration_uds_path = "/tmp/spire-registration.sock"
+	trust_domain = "domain.test"
+	data_dir = "/opt/spire/data/server"
+	log_level = "DEBUG"
+    ca_ttl = "90s"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafA/server/server.conf:8]: weird indentation

File: test/integration/suites/nested-rotation/leafA/server/server.conf [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        }
+	}
+	KeyManager "memory" {
+		plugin_data = {}
+	}
+ 	UpstreamAuthority "spire" {
+ 	    plugin_data = {
+ 	        server_address      = "intermediateA-server"
+ 	        server_port         = 8081
+            workload_api_socket = "/opt/spire/sockets/workload_api.sock"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafA/server/server.conf:31]: indentation

File: test/integration/suites/nested-rotation/leafA/server/server.conf [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+plugins {
+	DataStore "sql" {
+		plugin_data {
+			database_type = "sqlite3"
+			connection_string = "/opt/spire/data/server/datastore.sqlite3"
+		}
+	}
+	NodeAttestor "x509pop" {
+		plugin_data {
+            ca_bundle_path = "/opt/spire/conf/server/agent-cacert.pem"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafA/server/server.conf:21]: indentation

File: test/integration/suites/nested-rotation/leafB/server/server.conf [hunk 8]
```diff
@@ -0,0 +1,34 @@
+server {
+	bind_address = "0.0.0.0"
+	bind_port = "8081"
+	registration_uds_path = "/tmp/spire-registration.sock"
+	trust_domain = "domain.test"
+	data_dir = "/opt/spire/data/server"
+	log_level = "DEBUG"
+    ca_ttl = "90s"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafB/server/server.conf:8]: indentation

File: test/integration/suites/nested-rotation/leafB/server/server.conf [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+plugins {
+	DataStore "sql" {
+		plugin_data {
+			database_type = "sqlite3"
+			connection_string = "/opt/spire/data/server/datastore.sqlite3"
+		}
+	}
+	NodeAttestor "x509pop" {
+		plugin_data {
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafB/server/server.conf:20]: indentation

File: test/integration/suites/nested-rotation/leafB/server/server.conf [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        }
+	}
+	KeyManager "memory" {
+		plugin_data = {}
+	}
+ 	UpstreamAuthority "spire" {
+ 	    plugin_data = {
+ 	        server_address      = "intermediateB-server"
+ 	        server_port         = 8081
+            workload_api_socket = "/opt/spire/sockets/workload_api.sock"
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/leafB/server/server.conf:31]: indentation

File: test/integration/suites/nested-rotation/root/server/server.conf [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+plugins {
+	DataStore "sql" {
+		plugin_data {
+			database_type = "sqlite3"
+			connection_string = "/opt/spire/data/server/datastore.sqlite3"
+		}
+	}
+	NodeAttestor "x509pop" {
+		plugin_data {
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/root/server/server.conf:20]: indentation

File: test/integration/suites/nested-rotation/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,4 @@
+FROM spire-agent:latest-local AS openssl-agent-mashup
```
Comment (azdagron) [inline test/integration/suites/nested-rotation/Dockerfile:1]: nit: maybe something more generic?
```suggestion
FROM spire-agent:latest-local AS nested-agent
```

File: test/integration/common [hunk 13]
```diff
@@ -60,3 +60,20 @@ fingerprint() {
 	# `openssl sha1` on macOS and linux.
 	openssl x509 -in "$1" -outform DER | openssl sha1 -r | awk '{print $1}'
 }
+
+checkSyncedEntry() {
```
[the remark this answers is not in the document]
Comment (MarcosDY) [reply test/integration/common:64]: done

File: test/integration/suites/nested-rotation/09-create-workload-entries [hunk 14]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+log-debug "creating intermediatieA workload registration entry..."
```
[the remark this answers is not in the document]
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/09-create-workload-entries:3]: done

File: test/integration/suites/nested-rotation/README.md (hunk 1 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/README.md:26]: done

File: test/integration/suites/nested-rotation/intermediateA/server/server.conf (hunk 2 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/intermediateA/server/server.conf:31]: there was tabs with spaces, it is solved now

File: test/integration/suites/nested-rotation/intermediateB/server/server.conf (hunk 4 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/intermediateB/server/server.conf:31]: done

File: test/integration/suites/nested-rotation/leafA/server/server.conf (hunk 5 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafA/server/server.conf:8]: done

File: test/integration/suites/nested-rotation/leafA/server/server.conf (hunk 6 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafA/server/server.conf:31]: done

File: test/integration/suites/nested-rotation/leafA/server/server.conf (hunk 7 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafA/server/server.conf:21]: done

File: test/integration/suites/nested-rotation/leafB/server/server.conf (hunk 8 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafB/server/server.conf:8]: done

File: test/integration/suites/nested-rotation/leafB/server/server.conf (hunk 9 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafB/server/server.conf:20]: done

File: test/integration/suites/nested-rotation/leafB/server/server.conf (hunk 10 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/leafB/server/server.conf:31]: done

File: test/integration/suites/nested-rotation/intermediateA/agent/agent.conf (hunk 3 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/intermediateA/agent/agent.conf:29]: ```root-agent_1            | time="2020-04-14T23:19:03Z" level=error msg="Failed to collect all selectors for PID" error="workload attestor \"docker\" failed: rpc error: code = Unknown desc = Error response from daemon: client version 1.41 is too new. Maximum supported API version is 1.40" pid=27 subsystem_name=workload_api```

File: test/integration/suites/nested-rotation/root/server/server.conf (hunk 11 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/root/server/server.conf:20]: done

File: test/integration/suites/nested-rotation/Dockerfile (hunk 12 above)
Comment (MarcosDY) [reply test/integration/suites/nested-rotation/Dockerfile:1]: done

**Pull request merged**